Sprite slice scaling and tiling

ManevilleF opened the pull request:
# Objective

Implement sprite tiling and [9 slice scaling](https://en.wikipedia.org/wiki/9-slice_scaling) for `bevy_sprite`.
Allowing slice scaling and texture tiling.

Basic scaling vs 9 slice scaling:

![Traditional_scaling_vs_9-slice_scaling](https://user-images.githubusercontent.com/26703856/177335801-27f6fa27-c569-4ce6-b0e6-4f54e8f4e80a.svg)

Slicing example:

<img width="481" alt="Screenshot 2022-07-05 at 15 05 49" src="https://user-images.githubusercontent.com/26703856/177336112-9e961af0-c0af-4197-aec9-430c1170a79d.png">

Tiling example:

<img width="1283" alt="Screenshot 2022-07-06 at 14 35 34" src="https://user-images.githubusercontent.com/26703856/177552197-01ae4b40-9050-4339-9e71-f346749448a1.png">

# Solution

- `Sprite` now has a `draw_mode` field storing a `SpriteDrawMode` enum with three variants:
  - `Simple` 
  - `Tiled` to have sprites tile horizontally and/or vertically
  - `Sliced` allowing 9 slicing the texture and optionally tile some sections with a `Textureslicer`.
- I updated the `bevy_sprite` extraction stage to extract potentially multiple textures instead of one, depending on the `draw_mode`
- I added two examples showcasing the slicing and tiling feature.

# TODO list

- [x] slicing
- [x] scale factor limit
- [x] Slice Tiling:
  - [x]  Basic stretch mode
  - [x] Center tiling
  - [x] Sides tiling
- [x] Sprite tiling without slices

# Blocking
- [ ] ~~Requires https://github.com/bevyengine/bevy/pull/5247 to be merged to move the tile and slice computations in the `prepare` stage of the sprite render pipeline~~
- [ ] https://github.com/bevyengine/bevy/pull/6621
- [ ] Once #5103  is merged I can enable tiling and slicing for texture sheets.
- [ ] Once #5070 is merged I can add this to `bevy_ui`

# To discuss

I want to move the slicer to be an `Asset` and avoid to compute the slices every frame, like for `TextureAtlas` but the slicing depends on the `Sprite::custom_size` and therefore needs to be updated every frame or on `Changed<Sprite>` detection

ManevilleF commented:
Now I'm not sure if I can put the slicer in an `Asset`, I can compute the `Rect` values from a given image size and store it but the slices must be computed every frame since the sprite can be *scaled*  or *resized*.

@mockersf how did you do it in https://github.com/vleue/bevy_ninepatch ? I looked at the code after you gave me the link but I don't understand the logic

alice-i-cecile commented on Cargo.toml:
```diff
@@ -200,6 +200,16 @@ description = "Renders a sprite"
+
+[package.metadata.example.sprite_slice]
+name = "Sprite Slice"
+description = "Showcases sprite slicing"
```
```suggestion
description = "Showcases slicing sprites into sections that can be scaled independently via the 9-patch technique"
```

alice-i-cecile replied:
Try to avoid saying the same thing in the same words :) If you can explain it differently you may help explain the concept to different audience members.

alice-i-cecile commented on crates/bevy_sprite/src/rect.rs:
```diff
@@ -12,6 +12,20 @@ pub struct Rect {
 }
 
+/// Struct defining a [`Sprite`](crate::Sprite) border with padding values
+#[repr(C)]
```
Why do we need this?

ManevilleF replied:
Copied and pasted from `Rect`, I'll remove it

alice-i-cecile commented on crates/bevy_sprite/src/rect.rs:
```diff
@@ -12,6 +12,20 @@ pub struct Rect {
+/// Struct defining a [`Sprite`](crate::Sprite) border with padding values
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug, Reflect)]
+pub struct BorderRect {
```
This duplicates the `Rect` type, which is re-exported from `taffy`.

ManevilleF replied:
So should `bevy_sprite` include `taffy` as a dependency?

ManevilleF replied:
So should bevy_sprite include taffy as a dependency?

alice-i-cecile replied:
I'm not sure. I dislike that as it's not a light dependency.

I think we should keep this as it is for now, but probably open an issue once this is merged to think about how we can reduce tech debt here.

ManevilleF replied:
Now that I think about it I could use a `Rect` directly, with the `min/max` values. Adding a constructor from padding values could do the trick but it sacrifices ergonomics a bit

alice-i-cecile commented on crates/bevy_sprite/src/render/mod.rs:
```diff
@@ -233,23 +240,49 @@ pub fn extract_sprites(
-            image_handle_id: handle.id,
-            anchor: sprite.anchor.as_vec(),
-        });
+        if let Some(slicer) = slicer {
```
I don't love introducing more complexity + a branch into a hot-path system like this. I'd be tempted to benchmark this as one system vs two.

ManevilleF replied:
https://github.com/bevyengine/bevy/pull/5213#discussion_r914906314

would remove the extra component

alice-i-cecile commented on crates/bevy_sprite/src/texture_slice.rs:
```diff
@@ -0,0 +1,294 @@
+}
+
+impl TextureSlice {
+    pub fn tiled(self, stretch_value: f32, (tile_x, tile_y): (bool, bool)) -> Vec<Self> {
```
Doc strings :)

alice-i-cecile commented on crates/bevy_sprite/src/texture_slice.rs:
```diff
@@ -0,0 +1,294 @@
+
+/// Defines how a texture slice scales when resized
+#[derive(Debug, Clone, Default, Reflect)]
+pub enum SliceScaleMode {
```
Yes please! I like that design much better.

ManevilleF commented:
@alice-i-cecile  I changed the logic to enable sprite tiling even without slicing, using an enum in `Sprite::draw_mode` and removing `TextureSlicer` from being a component.

I think this is better.

But now that I changed a lot of stuff in the extraction stage of the sprite render pipeline, benchmarking is definetely required

ManevilleF commented:
`many_sprites` stress test:

On this MR:
```
2022-07-07T09:27:50.191875Z  INFO many_sprites: Sprites: 102400
2022-07-07T09:27:50.191926Z  INFO bevy diagnostic: frame_time                      :    0.058695s (avg 0.060466s)
2022-07-07T09:27:50.191937Z  INFO bevy diagnostic: fps                             :   17.037371  (avg 16.569809)
2022-07-07T09:27:50.191941Z  INFO bevy diagnostic: frame_count                     : 403.000000
```
On Main:
```
2022-07-07T09:29:16.821865Z  INFO many_sprites: Sprites: 102400
2022-07-07T09:29:16.821931Z  INFO bevy diagnostic: frame_time                      :    0.052531s (avg 0.050298s)
2022-07-07T09:29:16.821944Z  INFO bevy diagnostic: fps                             :   19.036325  (avg 19.903592)
2022-07-07T09:29:16.821949Z  INFO bevy diagnostic: frame_count                     : 415.000000
```

So there is some performance regression for regular sprites. I'm going to investigate this

ManevilleF commented:
The big issue is accessing the `Image` from the handle in the extraction stage. I removed this step for `SpriteDrawMode::Simple` sprites,  meaning no performance regressions.

Now this is still required for `Sliced` and `Tiled` draw modes. A more optimized solution would be to retrieve the image size lazily on the `Added<Sprite>` event or to ask the user to give the image dimensions.
I'm trying to think about a more ergonomic way to handle this, since the image dimensions are accessed again in the `prepare` stage of the pipeline.

superdump commented on crates/bevy_sprite/src/render/mod.rs:
```diff
@@ -237,19 +238,68 @@ pub fn extract_sprites(
+        };
+
+        let slices = match &sprite.draw_mode {
+            SpriteDrawMode::Sliced(slicer) => slicer.compute_slices(
```
Note though that if it doesn't take any significant amount of time for a lot of such entities (i.e. doesn't add more than say 0.1ms compared to as many entities that just use simple sprites) but it would take more time in a separate `prepare` stage system due to having to iterate over them all again, then it can be OK to do it here. So basically profile and present results.

superdump replied:
Having now seen the slicer code, I suggest moving that to the `prepare` stage.

ManevilleF replied:
I'll move every computation in `prepare` once https://github.com/bevyengine/bevy/pull/5247 is done

superdump commented on crates/bevy_sprite/src/texture_slice.rs:
```diff
@@ -0,0 +1,329 @@
+use bevy_reflect::Reflect;
+
+/// Slices a texture using the **9-slicing** technique. This allows to reuse an image at various sizes
+/// without needing to prepare multiple assets. The associated texture will be split tinto nine portions,
```
```suggestion
/// without needing to prepare multiple assets. The associated texture will be split into nine portions,
```

superdump commented on crates/bevy_sprite/src/texture_slice.rs:
```diff
@@ -0,0 +1,329 @@
+    pub border: BorderRect,
+    /// Defines how the center part of the 9 slices will scale
+    pub center_scale_mode: SliceScaleMode,
+    /// Defines how the 4 side part of the 9 slices will scale
```
```suggestion
    /// Defines how the 4 side parts of the 9 slices will scale
```

superdump commented on crates/bevy_sprite/src/texture_slice.rs:
```diff
@@ -0,0 +1,329 @@
+    pub sides_scale_mode: SliceScaleMode,
+    /// Defines the maximum scale of the 4 corner slices (default to `1.0`)
+    ///
+    /// Note: the value will be clamped to `0.001` if lower
```
How come?

ManevilleF commented:
Closing in favor of #10588
